dhis-2/dhis-services/dhis-service-field-filtering/src/main/java/org/hisp/dhis/fieldfiltering/FieldPathHelper.java

```diff
@@ -0,0 +1,314 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.fieldfiltering;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import lombok.RequiredArgsConstructor;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.schema.Property;
+import org.hisp.dhis.schema.PropertyType;
+import org.hisp.dhis.schema.Schema;
+import org.hisp.dhis.schema.SchemaService;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Morten Olav Hansen
+ */
+@Component
+@RequiredArgsConstructor
+public class FieldPathHelper
+{
+    public static final String PRESET_ALL = "all";
+
+    public static final String PRESET_OWNER = "owner";
+
+    private final SchemaService schemaService;
+
+    public void apply( List<FieldPath> fieldPaths, Class<?> rootKlass )
+    {
+        if ( rootKlass == null || fieldPaths.isEmpty() )
+        {
+            return;
+        }
+
+        List<FieldPath> presets = fieldPaths.stream().filter( FieldPath::isPreset ).collect( Collectors.toList() );
+        List<FieldPath> exclusions = fieldPaths.stream().filter( FieldPath::isExclude ).collect( Collectors.toList() );
+
+        fieldPaths.removeIf( FieldPath::isPreset );
+        fieldPaths.removeIf( FieldPath::isExclude );
+
+        Map<String, FieldPath> fieldPathMap = getFieldPathMap( fieldPaths );
+
+        applyProperties( fieldPathMap.values(), rootKlass );
+        applyPresets( presets, fieldPathMap, rootKlass );
+
+        calculatePathCount( fieldPathMap.values() ).forEach( ( k, v ) -> {
+            if ( v > 1L )
+            {
+                return;
+            }
+
+            applyDefaults( fieldPathMap.get( k ), rootKlass, fieldPathMap );
+        } );
+
+        applyExclusions( exclusions, fieldPathMap );
+
+        fieldPaths.clear();
+        fieldPaths.addAll( fieldPathMap.values() );
```

Comment:
Makes me think why not reconstruct a collection instead?

Reply:
This map is used to easily remove excluded properties